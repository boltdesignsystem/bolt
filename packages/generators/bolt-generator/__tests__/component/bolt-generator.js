const shell = require('shelljs');
const fs = require('fs');
const internalTasks = require('@bolt/build-tools/tasks/internal-tasks');
const dir = require('node-dir');

const dirs = {
  global: 'packages/generators/tmp',
  js: 'test.js',
  scss: 'test.scss',
  test: 'test.js',
  package: 'package.json',
  readme: 'README.md',
  schema: 'test.schema.js',
  boltrc: '.boltrc.js',
};

dirs.component = `${dirs.global}/packages/components/bolt-test`;
dirs.src = `${dirs.component}/src`;
dirs.testing = `${dirs.component}/__tests__`;
dirs.patternLab = `${dirs.global}/docs-site/src/pages/pattern-lab/_patterns/40-components/test/00-test-docs.twig`;

describe('Bolt component generator', () => {
  beforeAll(async () => {
    shell.exec(`rm -rf ${dirs.global}`);
    shell.exec('yarn cc "Test" "n" "Test Description"');
  });

  afterAll(async () => {
    await internalTasks.clean(dirs.global);
  });

  test('pattern lab file exist', async () => {
    const results = fs.readFileSync(dirs.patternLab, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('JS file exist', async () => {
    const results = fs.readFileSync(`${dirs.src}/${dirs.js}`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('SCSS file exist', async () => {
    const results = fs.readFileSync(`${dirs.src}/${dirs.scss}`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('tests file exist', async () => {
    const results = fs.readFileSync(`${dirs.testing}/${dirs.test}`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('index JS file exist', async () => {
    const results = fs.readFileSync(`${dirs.component}/index.js`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('index SCSS file exist', async () => {
    const results = fs.readFileSync(`${dirs.component}/index.scss`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('package.json file exist', async () => {
    const results = fs.readFileSync(
      `${dirs.component}/${dirs.package}`,
      'utf8',
    );

    expect(results).toMatchSnapshot();
  });

  test('readme file exist', async () => {
    const results = fs.readFileSync(`${dirs.component}/${dirs.readme}`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('schema file exist', async () => {
    const results = fs.readFileSync(`${dirs.component}/${dirs.schema}`, 'utf8');

    expect(results).toMatchSnapshot();
  });

  test('bolt package.json file is updated', async () => {
    const results = fs.readFileSync(`${dirs.global}/${dirs.package}`, 'utf8');

    expect(results.includes('@bolt/components-test')).toBe(true);
  });

  test('.boltrc.js is updated', async () => {
    const results = fs.readFileSync(`${dirs.global}/${dirs.boltrc}`, 'utf8');

    expect(results.includes('@bolt/components-test')).toBe(true);
  });

  test('folder structure of component package is correct', async () => {
    const results = await dir.files(dirs.global, { sync: true });

    expect(results).toMatchSnapshot();
  });
});
