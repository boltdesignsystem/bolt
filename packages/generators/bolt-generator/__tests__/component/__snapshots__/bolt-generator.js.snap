// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bolt component generator JS file exist 1`] = `
"export class BoltTest {
  constructor(el) {
    if (!el) return;
    this.el = el;
    this.init();
  }

  init() {
    console.log('This is a Test component.');
  }
}
"
`;

exports[`Bolt component generator SCSS file exist 1`] = `
"/* ------------------------------------ *\\\\
   Test
\\\\* ------------------------------------ */

@import '@bolt/core-v3.x';

// Register Custom Block Element
@include bolt-custom-element('bolt-test', block, medium);

/**
 * Long-form comment.
 *
 * This spans multiple lines and is also constrained to no longer than 80
 * characters in width.
 *
 * 1. Provide line-comments like this.
 */
.c-bolt-test {
  // Styles go here. /* [1] */
  display: block;
}
"
`;

exports[`Bolt component generator folder structure of component package is correct 1`] = `
Array [
  "packages/generators/bolt-generator/__tests__/component/_tmp/.boltrc.js",
  "packages/generators/bolt-generator/__tests__/component/_tmp/docs-site/src/pages/pattern-lab/_patterns/40-components/test/00-test-docs.twig",
  "packages/generators/bolt-generator/__tests__/component/_tmp/docs-site/src/pages/pattern-lab/_patterns/40-components/test/05-test.twig",
  "packages/generators/bolt-generator/__tests__/component/_tmp/package.json",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/README.md",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/__tests__/test.js",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/index.js",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/index.scss",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/package.json",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/src/test.js",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/src/test.scss",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/src/test.twig",
  "packages/generators/bolt-generator/__tests__/component/_tmp/packages/components/bolt-test/test.schema.js",
]
`;

exports[`Bolt component generator index JS file exist 1`] = `
"const tests = document.querySelectorAll('.c-bolt-test');

if (tests.length) {
  import(/* webpackChunkName: 'bolt-test' */ './src/test').then(
    ({ BoltTest }) => {
      tests.forEach(el => {
        const testComponent = new BoltTest(el);
      });
    },
  );
}
"
`;

exports[`Bolt component generator index SCSS file exist 1`] = `
"@import 'src/test.scss';
"
`;

exports[`Bolt component generator package.json file exist 1`] = `
"{
  \\"name\\": \\"@bolt/components-test\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"Test Description\\",
  \\"keywords\\": [\\"bolt design system\\"],
  \\"homepage\\": \\"https://boltdesignsystem.com\\",
  \\"bugs\\": {
    \\"url\\": \\"https://github.com/bolt-design-system/bolt/issues\\"
  },
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"https://github.com/bolt-design-system/bolt/tree/master/packages/components/bolt-test\\"
  },
  \\"license\\": \\"MIT\\",
  \\"author\\": \\"Test User <test@example.org>\\",
  \\"main\\": \\"index.js\\",
  \\"style\\": \\"index.scss\\",
  \\"dependencies\\": {
    \\"@bolt/core-v3.x\\": \\"^0.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\"
  },
  \\"schema\\": \\"test.schema.js\\",
  \\"twig\\": \\"src/test.twig\\"
}
"
`;

exports[`Bolt component generator pattern lab file exist 1`] = `
"{% set usage %}{% verbatim %}
{% include '@bolt-components-test/test.twig' with {
  content: 'This is a test component.'
} only %}
{% endverbatim %}{% endset %}

{% include '@utils/docs.twig' with {
  componentName: 'test',
  usage: usage
} only %}
"
`;

exports[`Bolt component generator readme file exist 1`] = `
"Test Description

###### Install via NPM

\`\`\`
npm install @bolt/components-test
\`\`\`
"
`;

exports[`Bolt component generator schema file exist 1`] = `
"module.exports = {
  $schema: 'http://json-schema.org/draft-04/schema#',
  title: 'Test',
  type: 'object',
  properties: {
    attributes: {
      type: 'object',
      description:
        'A Drupal attributes object. Applies extra HTML attributes to the &lt;bolt-test&gt; tag.',
    },
    title: {
      type: 'any',
      description: 'Title of Example.',
    },
    content: {
      type: 'any',
      description: 'The main Example content.',
    },
    disabled: {
      type: 'boolean',
      description:
        'Default \`disabled\` prop supported globally by most Bolt components.',
      default: false,
    },
  },
};
"
`;

exports[`Bolt component generator tests file exist 1`] = `
"// Refer to https://github.com/boltdesignsystem/bolt/wiki/Jest-Test:-Example-Jest-Test for more testing examples
import { render, stopServer } from '../../../testing/testing-helpers';
// import schema from '../test.schema';
// const { [PROP KEY], [PROP KEY]... } = schema.properties;
let page, fixtures;

afterAll(async () => {
  await stopServer();
  await page.close();
});

beforeEach(async () => {
  await page.evaluate(() => {
    document.body.innerHTML = '';
  });
});

beforeAll(async () => {
  page = await global.__BROWSER__.newPage();
  await page.goto('http://127.0.0.1:4444/', {
    timeout: 0,
  });

  const defaultData = {
    content: 'This is a test',
  };

  fixtures = {
    defaultData,
  };
});

describe('Bolt Test component', () => {
  test('default', async () => {
    const results = await render(
      '@bolt-components-test/test.twig',
      {
        ...fixtures.defaultData,
      },
    );

    await expect(results.ok).toBe(true);
    await expect(results.html).toMatchSnapshot();
  });
});

// describe('Bolt Test prop -', () => {
//   // Target each of the schema keys with the following pattern
//   [PROP KEY].enum.forEach(async (option) => {
//     test(\`[PROP KEY] items: \${option}\`, async () => {
//       const results = await render(
//         '@bolt-components-test/test.twig',
//         {
//           ...fixtures.defaultData,
//           [PROP KEY]: option,
//         }
//       );

//       await expect(results.ok).toBe(true);
//       await expect(results.html).toMatchSnapshot();
//     });
//   });
// });
"
`;
